/******************************************************************************
Copyright (c) 2021, Farbod Farshidian. All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

 * Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

 * Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

 * Neither the name of the copyright holder nor the names of its
  contributors may be used to endorse or promote products derived from
  this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
******************************************************************************/

#include <ocs2_centroidal_model/CentroidalModelPinocchioMapping.h>
#include <CatModel2_v2_interface/LeggedInterface.h>
#include <ocs2_pinocchio_interface/PinocchioEndEffectorKinematics.h>
#include <ocs2_ros_interfaces/mrt/MRT_ROS_Dummy_Loop.h>
#include <ocs2_ros_interfaces/mrt/MRT_ROS_Interface.h>

#include "CatModel2_v2_dummy/visualization/CatModel2Visualizer.h"
#include "rclcpp/rclcpp.hpp"

using namespace ocs2;
using namespace legged_robot;

int main(int argc, char **argv) {
    const std::string robotName = "legged_robot";

    // Initialize ros node
    rclcpp::init(argc, argv);
    rclcpp::Node::SharedPtr node = rclcpp::Node::make_shared(
        robotName + "_mrt",
        rclcpp::NodeOptions()
        .allow_undeclared_parameters(true)
        .automatically_declare_parameters_from_overrides(true));
    // Get node parameters
    const std::string taskFile = node->get_parameter("taskFile").as_string();
    const std::string urdfFile = node->get_parameter("urdfFile").as_string();
    const std::string referenceFile =
            node->get_parameter("referenceFile").as_string();

    // Robot interface
    LeggedInterface interface(taskFile, urdfFile, referenceFile);

    // MRT
    MRT_ROS_Interface mrt(robotName);
    mrt.initRollout(&interface.getRollout());
    mrt.launchNodes(node);

    // Visualization
    CentroidalModelPinocchioMapping pinocchioMapping(
        interface.getCentroidalModelInfo());
    PinocchioEndEffectorKinematics endEffectorKinematics(
        interface.getPinocchioInterface(), pinocchioMapping,
        interface.modelSettings().contactNames3DoF);
    auto leggedRobotVisualizer = std::make_shared<LeggedRobotVisualizer>(
        interface.getPinocchioInterface(), interface.getCentroidalModelInfo(),
        endEffectorKinematics, node);

    // Dummy legged robot
    MRT_ROS_Dummy_Loop leggedRobotDummySimulator(
        mrt, interface.mpcSettings().mrtDesiredFrequency_,
        interface.mpcSettings().mpcDesiredFrequency_);
    leggedRobotDummySimulator.subscribeObservers({leggedRobotVisualizer});

    // Initial state
    SystemObservation initObservation;
    initObservation.state = interface.getInitialState();
    initObservation.input =
            vector_t::Zero(interface.getCentroidalModelInfo().inputDim);
    initObservation.mode = ModeNumber::STANCE;

    // Initial command
    TargetTrajectories initTargetTrajectories({0.0}, {initObservation.state},
                                              {initObservation.input});

    // run dummy
    leggedRobotDummySimulator.run(initObservation, initTargetTrajectories);

    // Successful exit
    return 0;
}
